### PR TITLE
UVM: basic implementation of amap

### PR DESCRIPTION
**Pull request by pj1031999**
It's a first step for new virtual memory subsystem [1] in Mimiker.

amap describes an area of anonymous memory. It 's a upper layer of UVM’s two-layer mapping scheme.

* Add missing PAGE_SHIFT definition in vm_param for AArch64 & mips.
* Add definition of uvm_amap.
* Add interface for uvm_amap.
* Add definition of uvm_aref.
* Add simple test for uvm_amap.

It's based on NetBSD implementation [2] described by Cranor [3].

[1] #1073
[2] https://nxr.netbsd.org/xref/src/sys/uvm/uvm_amap.c
[3] https://chuck.cranor.org/p/diss.pdf

### CODE REVIEW

**Review by franciscozdo — CHANGES_REQUESTED** [review]
I would be happy to see assertions about held lock in functions that require it (at least in functions `*_nolock`). It could help in future.

**Comment by franciscozdo** [comment]
What do you think about moving tests into `sys/tests/uvm.c`. Then we will have all tests in one place.

File: sys/kern/uvm_amap.c [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+static POOL_DEFINE(P_AMAP, "uvm_amap", sizeof(uvm_amap_t));
+
+static inline int uvm_amap_slot(uvm_aref_t *aref, vaddr_t offset) {
+  assert(offset % PAGESIZE == 0);
+  return (int)(aref->ar_pageoff + (offset >> PAGE_SHIFT));
+}
+
+uvm_amap_t *uvm_amap_alloc(void) {
+  uvm_amap_t *amap = pool_alloc(P_AMAP, M_ZERO);
```
[the remark this answers is not in the document]
Comment (pj1031999) [reply sys/kern/uvm_amap.c:19]: Fixed

**Comment by pj1031999** [comment]
> What do you think about moving tests into `sys/tests/uvm.c`. Then we will have all tests in one place.

I'm not sure. From my point of view amap is a separate structure that can live without rest of UVM. But if you thing that one place for all tests is a better idea you can push that directly here. It's not a a big thing for me.

**Comment by pj1031999** [comment]
> I would be happy to see assertions about held lock in functions that require it (at least in functions `*_nolock`). It could help in future.

Hm... I'm not sure if it's necessary. That functions are private for amap and called only from their locked brothers now.
I've checked that and we have 5 `_nolock` functions in our code and only one has that assertion.
I can add them but I don't see a reason for that now.

**Review by franciscozdo — APPROVED** [review]

File: include/sys/vm_amap.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ * Field markings and the corresponding locks:
+ *  (@) vm_amap::am_lock
+ */
+struct vm_amap {
+  mtx_t am_lock;
+  int am_ref;     /* (@) reference counter */
+  int am_nslot;   /* (@) number of allocated slots */
+  int am_nused;   /* (@) number of used slots */
+  int *am_slot;   /* (@) slots of used anons - refers to am_bckptr */
+  int *am_bckptr; /* (@) stack of used anons - refers to am_anon & am_slots */
```
Comment (franciscozdo) [inline include/sys/vm_amap.h:32]: According to UVM functionality of am_slot and am_backptr should be swapped.

**Review by franciscozdo — CHANGES_REQUESTED** [review]

**Comment by franciscozdo** [comment]
@cahirwpz I think that it can be closed hence I will implement UVM like subsystem in #1379 and later PRs.

**Comment by cahirwpz** [comment]
Closing the PR as requested in https://github.com/cahirwpz/mimiker/pull/1072#issuecomment-1443985549

**Pull request closed without merging**